canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,864 @@
+    CustomData *data = reinterpret_cast<CustomData *>(custom_data);
+    if (pFragmentAck->ackType != FRAGMENT_ACK_TYPE_PERSISTED && pFragmentAck->ackType != FRAGMENT_ACK_TYPE_RECEIVED)
+    {
+        return STATUS_SUCCESS;
+    }
+
+    map<uint64_t, uint64_t>::iterator iter;
+    iter = data->timeOfNextKeyFrame->find(pFragmentAck->timestamp);
+    
+    uint64_t timeOfFragmentEndSent = data->timeOfNextKeyFrame->find(pFragmentAck->timestamp)->second;
+
+    if (timeOfFragmentEndSent > pFragmentAck->timestamp)
+    {
+        switch (pFragmentAck->ackType)
+        {
+            case FRAGMENT_ACK_TYPE_PERSISTED:
+            {
+                Aws::CloudWatch::Model::MetricDatum persistedAckLatencyDatum;
+                Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+                cwRequest.SetNamespace("KinesisVideoSDKCanary");
+
+                auto currentTimestamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+                auto persistedAckLatency = (currentTimestamp - timeOfFragmentEndSent); // [milliseconds]
+                pushMetric("PersistedAckLatency", persistedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, persistedAckLatencyDatum, data->pDimensionPerStream, cwRequest);
+                LOG_DEBUG("Persisted Ack Latency: " << persistedAckLatency);
+                if (data->pCanaryConfig->useAggMetrics)
+                {
+                    pushMetric("PersistedAckLatency", persistedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, persistedAckLatencyDatum, data->pAggregatedDimension, cwRequest);
+
+                }
+                data->pCWclient->PutMetricDataAsync(cwRequest, onPutMetricDataResponseReceivedHandler);
+                break;
+            }
+            case FRAGMENT_ACK_TYPE_RECEIVED:
+            {
+                Aws::CloudWatch::Model::MetricDatum receivedAckLatencyDatum;
+                Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+                cwRequest.SetNamespace("KinesisVideoSDKCanary");
+
+                auto currentTimestamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+                auto receivedAckLatency = (currentTimestamp - timeOfFragmentEndSent); // [milliseconds]
+                pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatencyDatum, data->pDimensionPerStream, cwRequest);
+                LOG_DEBUG("Received Ack Latency: " << receivedAckLatency);
+                if (data->pCanaryConfig->useAggMetrics)
+                {
+                    pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatencyDatum, data->pAggregatedDimension, cwRequest);
+                }
+                data->pCWclient->PutMetricDataAsync(cwRequest, onPutMetricDataResponseReceivedHandler);
```

Comment:
Same about putMetricDataAsync. Please fix this everywhere. We need per stream metrics as well and not just aggregated. Include the call as part of `pushMetric` API so that the `PutMetricDataAsync` is invoked only at one location. Makes code cleaner.